### ISSUE TITLE
update readme

### ISSUE THREAD

**Issue by stuart909** [dimaqq/awaitwhat#34]
```


    Hi group,

    I'm recently debugging a long-running asyncio program that appears to get stuck about once a week.

    The tools I've discovered so far are:

        high level: asyncio.all_tasks() + asyncio.Task.get_stack()
        low level: loop._selector._fd_to_key

    What's missing is the middle level, i.e. stack-like linkage of what is waiting for what. For a practical example, consider:

    async def leaf(): await somesocket.recv()
    async def baz(): await leaf()
    async def bar(): await baz()
    async def foo(): await bar()
    async def job(): await foo()
    async def work(): await asyncio.gather(..., job())
    async def main(): asyncio.run(work())

    The task stack will contain:

        main and body of work with line number
        job task with line number pointing to foo

    The file descriptor mapping, socket fd, loop._recv() and a Future.

    What's missing are connections foo->bar->baz->leaf. That is, I can't tell which task is waiting for what terminal Future.

    Is this problem solved in some way that I'm not aware of? Is there a library or external tool for this already?

    Perhaps, if I could get a list of all pending coroutines, I could figure out what's wrong.

    If no such API exists, I'm thinking of the following:

    async def foo():
        await bar()

    In [37]: dis.dis(foo)
      1           0 LOAD_GLOBAL              0 (bar)
                  2 CALL_FUNCTION            0
                  4 GET_AWAITABLE
                  6 LOAD_CONST               0 (None)
                  8 YIELD_FROM
                 10 POP_TOP
                 12 LOAD_CONST               0 (None)
                 14 RETURN_VALUE

    Starting from a pending task, I'd get it's coroutine and:

    Get the coroutine frame, and if current instruction is YIELD_FROM, then the reference to the awaitable should be on the top of the stack. If that reference points to a pending coroutine, I'd add that to the "forward trace" and repeat.

    At some point I'd reach an awaitable that's not a pending coroutine, which may be: another Task (I already got those), a low-level Future (can be looked up in event loop), an Event (tough luck, shoulda logged all Event's on creation) or a dozen other corner cases.

    What do y'all think of this approach?

    Thanks, D.



```
It appears that the initial issue you mentioned, experiencing a stuck process once a week, could be related to disk I/O problems. In my experience, regular disk checks, like the ones my NAS performs on Saturdays, can cause significant delays and even crashes if there are processes running that heavily rely on disk I/O.  This issue is also compounded by high user activity on the disks.

Regarding your async implementation, it sounds like you might benefit from considering threading if you need a higher level of parallelism. However, if you are tracking your tasks in a linear manner, creating a sequential flow of events through your async functions using await statements can work effectively. By properly ordering the dependencies with await, you ensure the tasks execute in the desired sequence.

In situations where you have a large number of tasks that can be completed mostly in parallel but have blocking dependencies, it's crucial to order those dependencies correctly using await statements, making sure the conditions are within the scope of your functions.

To troubleshoot potential bottlenecks and optimize your async program, consider incorporating clever insights for creating await points. Techniques such as using decorators to log await points, setting up breakpoints with a debugger, or implementing custom tracing functions can help identify and analyze the execution flow of your tasks.

By gaining a deeper understanding of the await points and their relationships, you can effectively troubleshoot and resolve any performance issues, including potential disk-related delays.

I'll be happy to share insights over a beer.

edit:
was curious if the readme was up to date and this got fixed

**Comment by dimaqq** [comment]
At the moment the project is dead, as low level details were changed in Python 3.10 and 3.11...
It's been a while, so I'm probably forgetting things, here's the summary:

1. Some C API was changed and code simply doesn't compile
1. Python no longer updates the coro "program counter" in the stack frame, as an optimisation for coro a calling b calling c, etc.; This was the case for sync statements too, but program counter was stamped on await, in case the coro got suspended. The latter no longer happens.

**Comment by dimaqq** [comment]
Perhaps snapshot analysis was never the right approach.

https://discuss.python.org/t/add-asyncio-monitoring-hooks/26426 may be better.
